### PR TITLE
Feature/log recording

### PR DESCRIPTION
**Pull request by marcoconti83**
# Reason for this pull request
- Log capturing was not reliable, as it was based on ASL cache which might be cleaned by the iOS (and it is also polluted, filling up quickly). 
- Log hooks were not working for Swift. This was due to two different implementations of the code logic in Obj-c and Swift
- There was some Objective-C++ code that, while probably very fast, was not easy to understand. E.g. tags could only be static strings that were passed around and never defined twice, as it was relying on pointer comparison*. I think having 2 languages in the project is already enough, no need to learn C++ to work on this.

(* even if probably the compiler would make them point to the same memory location if statically defined twice, but who knows).

# Changes
- Rewrote most of the logging infrastructure to share the same implementation (in Swift)!
- Added our own log cache 
- Removed C++ in favor of Swift

# Testing
- All the new functionalities are tested

# Known issues
This might slow down logging. In particular, the C++ implementation avoided string comparison and it was comparing pointers, which is much faster. Now we are creating a lot of intermediate strings.

### CODE REVIEW

File: Source/ZMSLogging.m [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#import "ZMCSystem/ZMCSystem-swift.h"
+
+void ZMLog(NSString *tag, char const * const filename, int linenumber, ZMLogLevel_t logLevel, NSString *format, ...)
+{
+    va_list args;
+    va_start(args, format);
+    NSString *output = [[NSString alloc] initWithFormat:format arguments:args];
+    va_end(args);
+    
+    [ZMSLog logWithLevel:logLevel :^NSString * _Nonnull{
```
Comment (daehn) [inline Source/ZMSLogging.m]: The missing parameter label looks a bit odd in Objective-C, is there a specific reason for it? If not we could add a @objc method name to the Swift declaration.

Comment (marcoconti83) [inline Source/ZMSLogging.m]: No specific reason. I actually didn't even know you could force omitting a parameter in obj-c. It just came out like that.

Comment (daehn) [inline Source/ZMSLogging.m]: On the other hand it looks like we will only use the ZMLog() function from Objective-C I guess

File: Source/CircularArray.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+/// for inserting an infinite amount of elements (at the cost of discarding the old ones)
+public struct CircularArray<T> {
+    
+    /// Max size
+    private let size : Int
+    
+    /// A circular array used to store all lines
+    /// Once it reaches the end (full), it starts to overwrite from the beginning
+    /// The same operation could be achieved by appending to the end and removing from
+    /// the frong, but this would have caused reallocating the array
```
Comment (marcoconti83) [inline Source/CircularArray.swift]: Fixed

File: Source/ZMSLogging.m (hunk 1 above)
Comment (marcoconti83) [inline Source/ZMSLogging.m]: Fixed

**Pull request merged**